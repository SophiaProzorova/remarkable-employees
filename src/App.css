/* Application */
.application{
    width: 80%;
    display: block;
    margin: auto;
}
/* App end */

/* Form new employer section */
section.application>button {    
    transform: translate(320%, 0%);
    transition: 0s;
}

form.newEmployer{
    width: 78%;
    margin: 50px auto;
    display: flex;
    flex-wrap: wrap;
    justify-content: space-around;
    visibility: visible;
    transition: 2s;
}
form.newEmployer.hidden{
    visibility: hidden;
}

form.newEmployer>* {
    display: flex;
    width: 50%;
    flex-direction: column;
    justify-content: flex-start;
    align-items: center;
}
label{
    display: flex;

}
label span{
    font-size: 0.75em;
    font-weight: 700;
    margin: 0.55em 0.5em 0.25em;
    opacity: 0;
    color: rgba(136, 132, 216, 0.75);
}
label.warning  span {
    opacity: 1;
}
.inputTitle {
    width: 20%;
}

input[type=submit], button{
    width: 20%;
    box-sizing: border-box;
    margin: 0 50px;
    height: 30px;
    justify-content: center;
    text-transform: capitalize;
    background: #e8e7f7;
    outline: none;
    border: 1px solid #e8e7f7;
    border-radius: 10px;
    margin-top: 1em;
}
input, select{
    width: 70%;
    border-radius: 10px;
    padding: 7px 4px;
}
select{
    width: 73.4%;
}

/* Form new employer section end */

/* Chart section */
.chart_section {
    display: flex;
    justify-content: space-evenly;
    align-items: center;
    margin: 3% auto;
}
.recharts-wrapper{
    border: 1px solid black;
}
.recharts-wrapper>svg{
    padding-top: 8%;
    height: 90%;
}
.recharts-wrapper:first-child::before{
    content: "Employees by Job Title";
    background-color: white;
    position: absolute;
    transform: translate(20%, -60%);
}
.recharts-wrapper:last-child::before{
    content: "Employees by Gender";
    background-color: white;
    position: absolute;
    transform: translate(20%, -60%);
}
/* Chart section end */

/* Table */
table{
    margin: 50px auto;
    width: 78%;
    height: 10%;
    overflow-y: scroll;
}

thead {
    background-color: rgba(136, 132, 216, 0.75);
}
tr:nth-child(even) {
    background-color: #a6a3e242;
}
td {
    padding-left: 15px;
}
/* Table end */