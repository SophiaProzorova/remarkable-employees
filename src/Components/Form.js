import { useForm } from "react-hook-form";

export default function Form(props) {
    const { isSubmitted, onSubmit } = props;
    const { register, formState: { errors }, reset, handleSubmit } = useForm();

    return <form
        aria-label='form'
        className={`newEmployer ${isSubmitted ? 'hidden' : ''}`}
        onSubmit={handleSubmit(onSubmit)}
    >
        <label className={errors.name?.type == "required" ? "warning" : ''}>
            <span>This is required</span>
            <input placeholder="Name"
                {...register("name", { required: true, maxLength: 20, pattern: /^[A-Za-z]+$/i })} />
        </label>

        <label className={errors.jobTitle?.type == "required" ? "warning" : ''}>
            <span>This is required</span>
            <input placeholder="Job title"
                {...register("jobTitle", { required: true, maxLength: 20, pattern: /^[A-Za-z]+$/i })} />
        </label>

        <label className={errors.tenure?.type == "required" ? "warning" : ''}>
            <span>This is required and value should be from 0 to 99</span>
            <input
                placeholder="Tenure"
                type="number"
                {...register("tenure", { required: true, min: 1, max: 99 })} />
        </label>

        <label className={errors.tenure?.type == "required" ? "warning" : ''}>
            <span>This is required </span>
            <select defaultValue='' {...register("gender", { required: true, validate: value => value != '' })}>
                <option disabled value=''>Gender</option>
                <option value="Female">Female</option>
                <option value="Male">Male</option>
            </select>
        </label>

        <input type="submit" id="submitForm" />
    </form>;
}