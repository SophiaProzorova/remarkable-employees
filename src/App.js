import jsonData from "./employees.json"
import { useState } from "react"
import Table from "./Components/Table"
import Chart from "./Components/Chart";
import Form from "./Components/Form";

function App() {

  const [employees, setEmployees] = useState(jsonData);
  const [isSubmitted, setIsSubmitted] = useState(true);

  const onSubmit = (data) => {
    setIsSubmitted(prev => !prev);
    // update employees info
    setEmployees(prevArray => {
      const temp = {
        id: employees.length,
        ...data
      }
      return [...prevArray, temp]
    })
  }

  return (
    <section className={"application"}>
      <h1>Corporate Employees</h1>

      {/* show button for adding new employer or form */}
      {isSubmitted ?
        <button onClick={() => { setIsSubmitted(prev => !prev); }}>Add employee</button>
        :
        <Form isSubmitted={isSubmitted} onSubmit={onSubmit}/>
      }

      <Table values={employees} />
      <Chart values={employees} />
    </section>
  );
}

export default App;
