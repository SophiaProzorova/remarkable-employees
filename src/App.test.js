import { render, screen, within } from '@testing-library/react';
import '@testing-library/jest-dom'
import userEvent from '@testing-library/user-event';
import App from './App';
import Form from './Components/Form';

describe('testing App', ()=>{
  it('renders no form', ()=>{
    render (<App />);
    expect(screen.queryByRole('form')).toBeNull();
  })
  it('renders form on click button', async ()=>{
    render (< App />);
    const newEmployerButton = screen.getByRole('button');
    userEvent.click(newEmployerButton);
    await expect(screen.getByRole('form')).toBeInTheDocument();
  })
  it('submit null values', ()=>{
   const {container} = render (< App />);
   const newEmployerButton = screen.getByRole('button');
   userEvent.click(newEmployerButton);
   const submitButton = screen.getByRole('button', {id:'submitForm'});
   userEvent.click(submitButton);
   expect(screen.queryByRole('form')).toBeInTheDocument();
  })

  it('App without form snapshot', ()=>{
    const {app} = render(<App />);
    expect(app).toMatchSnapshot();
  })
  it('App with form snapshot', ()=>{
    const {app} = render(<App />);
    const newEmployerButton = screen.getByRole('button');
    userEvent.click(newEmployerButton);
    expect(app).toMatchSnapshot();
  })


})
