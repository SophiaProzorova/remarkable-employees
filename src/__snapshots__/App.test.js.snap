// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing App App with form snapshot 1`] = `undefined`;

exports[`testing App App without form snapshot 1`] = `undefined`;
